go/stellar/acctbundle/acctbundle.go

```diff
@@ -0,0 +1,673 @@
+	if err != nil {
+		return res, resB64, err
+	}
+
+	// Secretbox
+	var nonce [libkb.NaclDHNonceSize]byte
+	nonce, err = libkb.RandomNaclDHNonce()
+	if err != nil {
+		return res, resB64, err
+	}
+	secbox := secretbox.Seal(nil, clearpack[:], &nonce, (*[libkb.NaclSecretBoxKeySize]byte)(&symmetricKey))
+
+	// Annotate
+	res = stellar1.EncryptedAccountBundle{
+		V:   1,
+		E:   secbox,
+		N:   nonce,
+		Gen: pukGen,
+	}
+
+	// Msgpack (outer) + b64
+	cipherpack, err := libkb.MsgpackEncode(res)
+	if err != nil {
+		return res, resB64, err
+	}
+	resB64 = base64.StdEncoding.EncodeToString(cipherpack)
+	return res, resB64, nil
+}
+
+// decodeParent decodes a base64-encoded encrypted parent bundle.
+func decodeParent(encryptedB64 string) (stellar1.EncryptedBundle, stellar1.Hash, error) {
+	cipherpack, err := base64.StdEncoding.DecodeString(encryptedB64)
+	if err != nil {
+		return stellar1.EncryptedBundle{}, stellar1.Hash{}, err
+	}
+	encHash := sha256.Sum256(cipherpack)
+	var enc stellar1.EncryptedBundle
+	if err = libkb.MsgpackDecode(&enc, cipherpack); err != nil {
+		return stellar1.EncryptedBundle{}, stellar1.Hash{}, err
+	}
+	return enc, encHash[:], nil
+}
+
+// unboxParent unboxes an encrypted parent bundle and decodes the visual portion of the bundle.
+// It validates the visible hash in the secret portion.
+func unboxParent(encBundle stellar1.EncryptedBundle, hash stellar1.Hash, visB64 string, puk libkb.PerUserKeySeed) (*stellar1.BundleRestricted, stellar1.BundleVersion, error) {
+	versioned, err := decryptParent(encBundle, puk)
+	if err != nil {
+		return nil, 0, err
+	}
+	version, err := versioned.Version()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	var bundleOut stellar1.BundleRestricted
+	switch version {
+	case stellar1.BundleVersion_V2:
+		bundleOut, err = unboxParentV2(versioned, visB64)
+		if err != nil {
+			return nil, 0, err
+		}
+	default:
+		return nil, 0, fmt.Errorf("unsupported parent bundle version: %d", version)
+	}
+
+	bundleOut.OwnHash = hash
+	if len(bundleOut.OwnHash) == 0 {
+		return nil, 0, errors.New("stellar account bundle missing own hash")
+	}
+
+	return &bundleOut, version, nil
+}
+
+func unboxParentV2(versioned stellar1.BundleSecretVersioned, visB64 string) (stellar1.BundleRestricted, error) {
+	var empty stellar1.BundleRestricted
+	visiblePack, err := base64.StdEncoding.DecodeString(visB64)
+	if err != nil {
+		return empty, err
+	}
+	visibleHash := sha256.Sum256(visiblePack)
+	secretV2 := versioned.V2()
+	if !hmac.Equal(visibleHash[:], secretV2.VisibleHash) {
+		return empty, errors.New("corrupted bundle: visible hash mismatch")
+	}
+	var visibleV2 stellar1.BundleVisibleV2
+	err = libkb.MsgpackDecode(&visibleV2, visiblePack)
+	if err != nil {
+		return empty, err
+	}
+	return merge(secretV2, visibleV2)
+}
+
+// decryptParent decrypts an encrypted parent bundle with the provided puk.
+func decryptParent(encBundle stellar1.EncryptedBundle, puk libkb.PerUserKeySeed) (stellar1.BundleSecretVersioned, error) {
```

Comment:
this func is a dup of `bundle.Decrypt` with different error messages

Reply:
which already has tests in boxer_test.go

Reply:
changed to use bundle.Decrypt